lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,370 @@
+			outs() << "RULE HEAD IFF: " << *rule_head_iff << "\n";
+
+			//construct new body
+			ExprVector new_body_exprs;
+			new_body_exprs.push_back(rule_head_iff);
+
+			//Record occurence of rel in body, for increasing the number of primes
+			std::map<Expr, int> relOccurrenceInBodyMap;
+
+			//For each predicate in the body, construct new version of predicate.
+			Expr rule_body = r.body();
+			ExprVector body_pred_apps;
+			get_all_pred_apps(rule_body, db, std::back_inserter(body_pred_apps));
+			for(ExprVector::iterator it = body_pred_apps.begin(); it != body_pred_apps.end(); ++it)
+			{
+				Expr rule_body_pred = *it;
+				assert(oldToNewPredMap.find(bind::fname(rule_body_pred)) != oldToNewPredMap.end());
+				Expr new_rule_body_rel = oldToNewPredMap.find(bind::fname(rule_body_pred))->second;
+
+				if(relOccurrenceInBodyMap.find(bind::fname(rule_body_pred)) == relOccurrenceInBodyMap.end())
+				{
+					relOccurrenceInBodyMap.insert(std::pair<Expr, int>(bind::fname(rule_body_pred), 1));
+				}
+				else
+				{
+					relOccurrenceInBodyMap[bind::fname(rule_body_pred)] += 1;
+				}
+
+				ExprVector new_rule_body_args;
+				//Push boolean variables into arguments of new predicate
+				for(std::map<Expr, int>::iterator it = predToBiNumMap.begin(); it != predToBiNumMap.end(); ++it)
+				{
+					if(it->first == bind::fname(rule_body_pred))
+					{
+						int occur_num = relOccurrenceInBodyMap[bind::fname(rule_body_pred)];
+						assert(occur_num >= 1);
+						int bool_var_index = it->second;
+						std::ostringstream oss;
+						oss << "b" << bool_var_index;
+						for(int i=0; i<occur_num - 1; i++)
+						{
+							oss << "\'";
+						}
+						Expr bi = bind::boolVar(mkTerm<std::string>(oss.str(), db.getExprFactory ()));
+						predAppToBoolVarMap.insert(std::pair<Expr, Expr>(rule_body_pred, bi));
+						new_rule_body_args.push_back(bi);
+					}
+					else
+					{
+						int bool_var_index = it->second;
+						std::ostringstream oss;
+						oss << "b" << bool_var_index;
+						Expr bi = bind::boolVar(mkTerm<std::string>(oss.str(), db.getExprFactory ()));
+						new_rule_body_args.push_back(bi);
+					}
+				}
+				Expr new_rule_body_pred = bind::fapp(new_rule_body_rel, new_rule_body_args);
+				new_body_exprs.push_back(new_rule_body_pred);
+			}
+			//for each predicate in the body, create iff
+			for(ExprVector::iterator it = body_pred_apps.begin(); it != body_pred_apps.end(); ++it)
+			{
+				Expr rule_body_pred = *it;
+				Expr rule_body_cand_app = fAppToCandApp(rule_body_pred);
+				Expr rule_body_bool_var = predAppToBoolVarMap[rule_body_pred];
+				Expr rule_body_iff = mk<IFF>(rule_body_bool_var, rule_body_cand_app);
+				new_body_exprs.push_back(rule_body_iff);
+			}
+			//Extract the constraints
+			Expr constraints = extractTransitionRelation(r, db);
+			new_body_exprs.push_back(constraints);
+			//Construct new body
+			Expr new_rule_body = mknary<AND>(new_body_exprs.begin(), new_body_exprs.end());
```

Comment:
Can `new_body_exprs` be empty? Use the other `mknary` that takes the base result as an extra argument.